orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/expressions/functions/ServiceKeyExpressionFunctionProvider.java

```diff
@@ -0,0 +1,103 @@
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.orca.pipeline.expressions.functions;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.node.TreeTraversingParser;
+import com.netflix.spinnaker.orca.ExecutionStatus;
+import com.netflix.spinnaker.orca.jackson.OrcaObjectMapper;
+import com.netflix.spinnaker.orca.pipeline.expressions.ExpressionFunctionProvider;
+import com.netflix.spinnaker.orca.pipeline.model.Execution;
+import com.netflix.spinnaker.orca.pipeline.model.Stage;
+import lombok.Data;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import org.springframework.stereotype.Component;
+
+import java.io.IOException;
+import java.util.*;
+import java.util.function.Predicate;
+
+@Component
+public class ServiceKeyExpressionFunctionProvider implements ExpressionFunctionProvider {
+  private static final String CREATE_SERVICE_KEY_STAGE_NAME = "createServiceKey";
+  private static final ObjectMapper objectMapper = OrcaObjectMapper.newInstance();
+
+  @Nullable
+  @Override
+  public String getNamespace() {
+    return null;
+  }
+
+  @NotNull
+  @Override
+  public Collection<FunctionDefinition> getFunctions() {
+    return Collections.singletonList(
+      new FunctionDefinition("serviceKey", Arrays.asList(
+        new FunctionParameter(Execution.class, "execution", "The execution to search for stages within"),
+        new FunctionParameter(String[].class, "ids", "A list of stage name or stage IDs to search")
+      ))
+    );
+  }
+
+  public static Map<String, Object> serviceKey(Execution execution, String... id) {
+    return execution.getStages()
+      .stream()
+      .filter(matchesServiceKeyStage(id))
+      .findFirst()
+      .map(stage -> {
+        Map<String, Object> serviceKeyDetails = new HashMap<>();
+
+        Optional.ofNullable(stage.getContext().get("kato.tasks"))
+          .ifPresent(k -> {
+            List<Map<String, Object>> katoTasks = (List<Map<String, Object>>) k;
+            try {
+              ServiceKeyKatoTask katoTask = objectMapper.readValue(
+                new TreeTraversingParser(objectMapper.valueToTree(katoTasks.get(0)), objectMapper),
+                ServiceKeyKatoTask.class);
+              serviceKeyDetails.putAll(katoTask.getResultObjects().get(0).getServiceKey());
+            } catch (IOException e) {
+            }
+          });
+
+          return serviceKeyDetails;
+        })
+      .orElse(Collections.emptyMap());
+  }
+
+  private static Predicate<Stage> matchesServiceKeyStage(String... id) {
+    List<String> idsOrNames = Arrays.asList(id);
```

Comment:
can/should be `final`

Reply:
IntelliJ complains:  `As private methods cannot be meaningfully overridden, declaring them final is redundant.`.  We've been suggested to follow IntelliJ's suggestions in the past.  It's also consistent with https://github.com/spinnaker/orca/blob/820148ef29850cdce320aa0af3cad484065f4989/orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/expressions/functions/DeployedServerGroupsExpressionFunctionProvider.java#L93.

Reply:
i meant the `idsOrNames`

Reply:
We've eliminated the variable in response to another of your suggestions.